db_master_deploy/deploy.sh

```diff
@@ -0,0 +1,331 @@
+    if [ "$rows" -lt 1000 ]; then jobs=1;increment=$rows; fi
+    echo "multithread copy $source_id to $target_id rows: $rows threads: $jobs rows/thread: $increment size: $size attached slaves: $attached_slaves" 
+
+    echo "drop indexes on $target_id"
+    (pg_dump -U pgkogis -h localhost --if-exists -c -t $source_schema.$source_table -s $source_db 2>/dev/null | egrep "\bDROP INDEX\b" | psql -d "$target_db" -U pgkogis -h localhost 2>/dev/null )
+
+    # populate array with foreign key constraints on target table
+    declare -A foreign_keys=( )
+    while IFS=$'\|' read -a Record; do
+        foreign_keys["${Record[0]}"]="${Record[1]}"
+    done < <(
+    psql -U pgkogis -h localhost --quiet --no-align  -t -c "
+    SELECT conname,
+        pg_catalog.pg_get_constraintdef(r.oid, true) as condef
+    FROM pg_catalog.pg_constraint r
+    WHERE r.conrelid = '$target_schema.$target_table'::regclass AND r.contype = 'f' ORDER BY 1;" $target_db 2>/dev/null
+    )
+
+    if [ "${#foreign_keys[@]}" -gt 0 ]
+    then
+        for i in "${!foreign_keys[@]}"
+        do
+            echo "DROP FOREIGN KEY CONSTRAINT $i FROM $target_id ..."
+            echo "ALTER TABLE IF EXISTS ONLY $target_schema.$target_table DROP CONSTRAINT IF EXISTS $i;" | psql -U pgkogis -h localhost -d $target_db &> /dev/null
+        done
+    fi
+
+    echo "truncate table $target_id"
+    ( psql -U pgkogis -h localhost -c "begin; TRUNCATE TABLE $target_schema.$target_table; commit;" -d "$target_db" )
+
+    (
+    for i in $(eval echo {1..$jobs})
+    do 
+        offset=$(echo "(($i-1)*$increment)" | bc)
+        if [ $((offset+$increment)) -gt $rows ]; then counter=$rows; else counter=$((offset+$increment));fi
+        echo "dumping $offset..$counter"
+        ( psql -U pgkogis -h localhost -qAt -d $source_db -c "COPY ( SELECT * FROM $source_schema.$source_table order by ${primary_keys:=1} asc offset $offset limit $increment ) TO STDOUT with csv" | psql -U pgkogis -h localhost -qAt -d $target_db -c "SET session_replication_role = replica; COPY $target_schema.$target_table from stdin with csv; SET session_replication_role = DEFAULT;" )& pids="$pids $!"
+    done;  
+    wait $pids 2> /dev/null
+    )
+
+    echo "create indexes on $target_id"
+    ( pg_dump -U pgkogis -h localhost --if-exists -c -t $source_schema.$source_table -s $source_db 2>/dev/null | egrep -i "\bcreate\b" | egrep -i "\bindex\b" | sed "s/^/set search_path = $source_schema, public, pg_catalog; /" | sed "s/'/\\\'/g" | xargs --max-procs=$jobs -I '{}' sh -c 'psql -U pgkogis -h localhost -d $@ -c "{}"' -- "$target_db" ) 
+
+    if [ "${#foreign_keys[@]}" -gt 0 ]
+    then
+        for i in "${!foreign_keys[@]}"
+        do
+            echo "CREATE FOREIGN KEY CONSTRAINT $i ON $target_id ..."
+            echo "ALTER TABLE ONLY $target_schema.$target_table ADD CONSTRAINT $i ${foreign_keys[$i]};" | psql -U pgkogis -h localhost -d $target_db &> /dev/null
+        done
+    fi
+}
```

Comment:
My limited knowledge can't really review this part of the code. I don't know enough about pq. Who could review this in detail?